capsules/extra/src/ethernet_tap.rs

```diff
@@ -0,0 +1,658 @@
+                            Ok(())
+                        }
+                    })
+                    .unwrap()
+            })
+            .unwrap_or(Err(ErrorCode::FAIL))
+    }
+}
+
+/// Userspace system call driver interface implementation
+impl<'a, E: EthernetAdapterDatapath<'a>> SyscallDriver for EthernetTapDriver<'a, E> {
+    fn command(
+        &self,
+        command_num: usize,
+        arg1: usize,
+        arg2: usize,
+        process_id: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            // Check if driver is installed
+            0 => CommandReturn::success(),
+
+            // Query process-specific RX stats
+            1 => self
+                .apps
+                .enter(process_id, |grant, _kernel_data| {
+                    CommandReturn::success_u32_u32_u32(
+                        grant.rx_frames,
+                        grant.rx_bytes,
+                        grant.rx_frames_dropped,
+                    )
+                })
+                .unwrap_or(CommandReturn::failure(ErrorCode::FAIL)),
+
+            // Query process-specific TX stats
+            2 => self
+                .apps
+                .enter(process_id, |grant, _kernel_data| {
+                    CommandReturn::success_u32_u32(grant.tx_frames, grant.tx_bytes)
+                })
+                .unwrap_or(CommandReturn::failure(ErrorCode::FAIL)),
+
+            // Transmit frame from the `allow_ro::TX_FRAME` buffer:
+            3 => match self.command_transmit_frame(process_id, arg1 as u16, arg2 as u32) {
+                Ok(()) => CommandReturn::success(),
+                Err(e) => CommandReturn::failure(e),
+            },
+
+            _ => CommandReturn::failure(ErrorCode::NOSUPPORT),
+        }
+    }
+
+    fn allocate_grant(&self, processid: ProcessId) -> Result<(), kernel::process::Error> {
+        self.apps.enter(processid, |_, _| {})
+    }
+}
+
+/// Callback client for the underlying [`EthernetAdapterDatapath`]:
+impl<'a, E: EthernetAdapterDatapath<'a>> EthernetAdapterDatapathClient
+    for EthernetTapDriver<'a, E>
+{
+    fn transmit_frame_done(
+        &self,
+        err: Result<(), ErrorCode>,
+        frame_buffer: &'static mut [u8],
+        len: u16,
+        transmission_identifier: usize,
+        timestamp: Option<u64>,
+    ) {
+        // Complete this transmission. This will reset both the process'
+        // `tx_pending` state, and the global `tx_state`:
+        self.complete_transmission(
+            err,
+            frame_buffer,
+            len,
+            transmission_identifier as u32,
+            timestamp,
+        );
+
+        // Now, check for any processes that have a pending transmission. We use
+        // `find` as a short-circuiting method to iterate over the grant regions
+        // and only schedule a single transmission:
+        self.tx_state
+            .map(|tx_state| {
+                self.apps.iter().for_each(|process_grant| {
```

Comment:
Just realized that this doesn't actually use `find()` as indicated by the comment, which could cause a panic trying to `take()` the transmit buffer more than once.

Reply:
Fixed!